### PR TITLE
Replace METHOD string with COMMAND_ID integer (to remove obvious strings)

### PR DESCRIPTION
**Pull request by OJ**
# Intro

image

# Associated PRs

* [Meterpreter](https://github.com/rapid7/metasploit-payloads/pull/395)
* [Mettle](https://github.com/rapid7/mettle/pull/196)

# Description

This may be another longer story. I'm sorry, but detail is important.

Among the many goals we have right now, one of them is to make Meterp less obvious. That is, as a binary on disk/in memory, and with behaviour both on and off the wire.

Meterp is full of "commands". When operators tell `meterpreter >` to do something, behind the scenes TLV packets are created that contain a "method", which basically represents the identifier of the command the packet represents. That identifier was a string, and not just any string, a string that stands out from the crowd.

For those that haven't seen them, examples include `stdapi_fs_file_copy`, `core_migrate` and `mimikatz_custom_command`.

While these don't matter so much on the wire once TLV packets are encrypted, they do matter in other scenarios, such as strings embedded in binaries. The presence of anything `stdapi_` as a string tends to indicate bad things as far as AV is concerned.

The primary goal of this PR is to change `TLV_TYPE_METHOD` (a string) to `TLV_TYPE_COMMAND_ID` (an integer). This means that the obvious strings are removed from binaries and aren't present on the wire any more. This should have some impact on the obviousness of the binaries that we're producing.

This does mean that the ABI has been broken again, sorry about that. While doing this work, I was forced to make a bunch of other changes and did some other tidying up as well. But on the whole this PR is summed up as "convert a string to an int".

# Changes

Hopefully I haven't missed anything in this list:

* `TLV_TYPE_METHOD` no longer exists, and has been replaced by `TLV_TYPE_COMMAND_ID` which is an integer.
* Components now have an identifier `EXTENSION_ID_*`. This serves a few purposes:
    * It allows us to identify an extension by a number (which for `$REASONS` must be a multiple of `1000`.
    * We can put all commands identifiers for that extension within that `1000` number range (as no extension will have more than `1000` commands).
* Command IDs are grouped based on extension as per above.
* There's a thing called the `ExtensionMapper` that is used by the configuration generator and the extension loader. This thing has the responsibility of getting hold of extension classes, as well as being able to provide the identity of an extension based on its name. You can give it `"stdapi"` and it'll give you `1000` for example.
* Enumeration of supported extension commands now behaves differently. Instead of asking "what commands do you have for this extension", we now ask "which commands fit within this range". This means that:
    * We don't need to have strings in memory that map to extensions.
    * All extensions have had those strings removed.
    * Asking for support of an extension requires just two numbers.
* Both loading of extensions and enumerating supported commands both result in an array of command identifier integers rather than strings.
* Configuration block has been changed a little in the `EXTINIT` functionality. Given that we don't have string identifiers for the extensions any more, we now us the ID of the extension as an integer. This means that stageless payloads look a little different too.
* All the Meterpreters have had to change as a result, given the ABI modification. Doing this was FUN :|
* While refactoring stuff, I noticed that `networkpug` was still a thing. I removed it after talking to @bcook-r7 about it, because the old linux Meterpreter (which no longer exists) was the only thing that used it. Bye bye networkpug.
* Various other code tidying and whatnot (including removal of whitespace in certain cases). Sorry, that makes this noisy, but I was there!
* I exposed the "session remove" feature from the powershell extension while doing stuff. I realise this should be in another PR, but I was there, so it's done. Sorry!

# Verification

This may be a little painful, and for that I am sorry. _ALL_ of the Meterps have changed, so full regression testing of stuff will be required. I'm sure I have covered most of the cases, but there's a good chance I've missed something. Please help me make sure I haven't made any mistakes.

- [ ] Pull the updates for `mettle` and `metasploit-paylodas`, compile/build/etc and have them all stashed in the appropriate folders under `$MSF/data`.
- [ ] Create sessions for each with all the different transports/architectures/etc.
- [ ] For each of the sessions, run regression tests to make sure that things don't break.
- [ ] Simple commands should work.
- [ ] Channels should work (this will require a bit of focus to make sure we didn't make a mistake).
- [ ] Run the post test modules.
- [ ] Validate that the newly exposed Powershell extension feature (session removal) works as intended.

# Things left to do

Thanks to Channels having a string-based identifier, there's still some evidence of stuff like `stdapi_` in the code. Once this PR has been landed, I'll look to fix that up as well. I didn't want it to include that in this PR given it's already so big.

# Thanks

Cheers as always to Brent and Caitlin for the support. Thanks to Tim for his help getting more going with the Android stuff. Thanks to Spencer for his original work on the Python side that I stole while doing this.

# Attention

Given the depth/breadth of these changes, I'm going to need lots of eyes. There are many of you out there that can help but specifically I need the following folks for their expertise in certain areas: @bcook-r7 @smcintyre-r7 @jmartin-r7 @timwr - changes in MSF might not be that interesting to all of you, but reviews of the stuff I've done in the payloads/mettle PRs would be greatly appreciated. DO YOUR WORST!

### CODE REVIEW

**Review by sempervictus — COMMENTED** [review]
Did first pass w/ OJ - tiny fixups nothing major, amazing stuff as always from the master of meterpreter.

**Comment by OJ** [comment]
Flattery will get you everywhere @sempervictus !

**Comment by OJ** [comment]
BTW I will fix up the payload cache sizes when we've got the payloads PRs landed.

**Comment by OJ** [comment]
Ah need to fix up the specs.

**Comment by OJ** [comment]
image

Technically, that's where it started. This was a complete rewrite though 🚑

**Comment by kernelsmith** [comment]
Everyone stop pretending like they know what's going on in this PR and just merge it. Worst case we revert it and send OJ a glitter bomb

**Comment by bwatters-r7** [comment]
I've not looked too deeply, but it seems like sanity tests should fail because we've changed the dispatching system, but sanity tests have passed?

**Comment by jmartin-r7** [comment]
Local verification run of sanity test shows the expected failures due to these transport changes.  I am investigating why automation did not flag as expected.

Original `a2k19` discussion had suggested making this backwards compatible with existing payloads by detecting protocol format on initial connection, was there a conscious decision to make this a breaking change for all existing deployed payloads?

It would be nice to at least allow `reverse` protocols to be detected and fallback to original TLV formats.  I suspect this is may be complex to support and accept if decision has been made to limit scenario is the wild.

This PR may reflect a significant enough change in the how Metasploit communicates to call for a version bump to `5.1.0`.

**Comment by OJ** [comment]
I do remember that discussion. Keeping both wasn't totally out of the question, but boy was the code messy. It just didn't feel like the right thing to do. Thing is, as far as breaking back compat goes this is no worse than the refactor/removal of delay loading, though I'll admit that was just on the Meterp side.

Obviously it's not my call, but I think this has to be done and we just need to go through the pain of dealing with the changes and making sure that people don't upgrade mid-engagement.

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

Sanity test should correctly report failure now, once dependent gems are released and spec/lock are bumped in this PR they should pass again.

**Comment by OJ** [comment]
Thanks @jmartin-r7 !

**Comment by OJ** [comment]
Picked up on a few issues so please refrain from landing at this point :) FIXIFICATING!

**Comment by OJ** [comment]
OK back on track.

**Comment by bwatters-r7** [comment]
@msjenkins-r7 test this please.

**Comment by bwatters-r7** [comment]
@msjenkins-r7 test this please.

**Comment by bwatters-r7** [comment]
Hooray!  It is failing as expected, now.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/android.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # -*- coding: binary -*-
 require 'rex/post/meterpreter'
+require 'rex/post/meterpreter/exensions/android/command_ids'
```
Comment (timwr) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/android.rb:3]: exensions -> extensions

**Comment by timwr** [comment]
Python meterpreter is working great on Linux:
```
msf5 exploit(multi/handler) > sessions

Active sessions
===============

  Id  Name  Type                      Information    Connection
  --  ----  ----                      -----------    ----------
  1         meterpreter python/linux  ---

msf5 exploit(multi/handler) > sessions 1
[*] Starting interaction with 1...

meterpreter > help

Core Commands
=============

    Command                   Description
    -------                   -----------
    ?                         Help menu
    background                Backgrounds the current session
    bg                        Alias for background
    bgkill                    Kills a background meterpreter script
    bglist                    Lists running background scripts
    bgrun                     Executes a meterpreter script as a background thread
    channel                   Displays information or control active channels
    close                     Closes a channel
    disable_unicode_encoding  Disables encoding of unicode strings
    enable_unicode_encoding   Enables encoding of unicode strings
    exit                      Terminate the meterpreter session
    get_timeouts              Get the current session timeout values
    guid                      Get the session GUID
    help                      Help menu
    info                      Displays information about a Post module
    irb                       Open an interactive Ruby shell on the current session
    load                      Load one or more meterpreter extensions
    machine_id                Get the MSF ID of the machine attached to the session
    migrate                   Migrate the server to another process
    pry                       Open the Pry debugger on the current session
    quit                      Terminate the meterpreter session
    read                      Reads data from a channel
    resource                  Run the commands stored in a file
    run                       Executes a meterpreter script or Post module
    secure                    (Re)Negotiate TLV packet encryption on the session
    sessions                  Quickly switch to another session
    set_timeouts              Set the current session timeout values
    sleep                     Force Meterpreter to go quiet, then re-establish session.
    transport                 Change the current transport mechanism
    use                       Deprecated alias for "load"
    uuid                      Get the UUID for the current session
    write                     Writes data to a channel


Stdapi: File system Commands
============================

    Command       Description
    -------       -----------
    cat           Read the contents of a file to the screen
    cd            Change directory
    checksum      Retrieve the checksum of a file
    chmod         Change the permissions of a file
    cp            Copy source to destination
    del           Delete the specified file
    dir           List files (alias for ls)
    download      Download a file or directory
    edit          Edit a file
    getlwd        Print local working directory
    getwd         Print working directory
    lcd           Change local working directory
    lls           List local files
    lpwd          Print local working directory
    ls            List files
    mkdir         Make directory
    mv            Move source to destination
    pwd           Print working directory
    rm            Delete the specified file
    rmdir         Remove directory
    search        Search for files
    upload        Upload a file or directory


Stdapi: Networking Commands
===========================

    Command       Description
    -------       -----------
    ifconfig      Display interfaces
    ipconfig      Display interfaces
    portfwd       Forward a local port to a remote service
    resolve       Resolve a set of host names on the target


Stdapi: System Commands
=======================

    Command       Description
    -------       -----------
    execute       Execute a command
    getenv        Get one or more environment variable values
    getpid        Get the current process identifier
    getuid        Get the user that the server is running as
    kill          Terminate a process
    localtime     Displays the target system local date and time
    pgrep         Filter processes by name
    pkill         Terminate processes by name
    ps            List running processes
    shell         Drop into a system command shell
    sysinfo       Gets information about the remote system, such as OS


Stdapi: Audio Output Commands
=============================

    Command       Description
    -------       -----------
    play          play a waveform audio file (.wav) on the target system

meterpreter > background
[*] Backgrounding session 1...
msf5 exploit(multi/handler) > loadpath test/modules
Loaded 35 modules:
    13 exploit modules
    12 auxiliary modules
    10 post modules
msf5 exploit(multi/handler) > use post/test/meterpreter
msf5 post(test/meterpreter) > set SESSION -1
SESSION => -1
msf5 post(test/meterpreter) > run

[*] Running against session -1
[*] Session type is meterpreter and platform is linux
[*] Session doesn't implement getpid, skipping test
[+] should return a list of processes
[+] should return a user id
[+] should return a sysinfo Hash
[+] should return the proper directory separator
[+] should return the current working directory
[+] should list files in the current directory
[+] should stat a directory
[+] should create and remove a dir
[+] should change directories
[+] should create and remove files
[+] should upload a file
[+] should move files
[+] should copy files
[+] should do md5 and sha1 of files
[*] Passed: 14; Failed: 0
[*] Post module execution completed
msf5 post(test/meterpreter) > use post/test/cmd_exec
msf5 post(test/cmd_exec) > set SESSION -1
SESSION => -1
msf5 post(test/cmd_exec) > run

[*] Running against session -1
[*] Session type is meterpreter and platform is linux
[+] should return the result of echo with single quotes
[+] should return the result of echo with double quotes
[+] should return the result of echo
[+] should return the full response after sleeping
[+] should return the full response after sleeping
[+] should return the result of echo 10 times
[*] Passed: 6; Failed: 0
[*] Post module execution completed

```

**Comment by timwr** [comment]
Some tests are failing on Java now.
```
msf5 exploit(multi/handler) > loadpath test/modules
Loaded 35 modules:
    13 exploit modules
    12 auxiliary modules
    10 post modules
msf5 exploit(multi/handler) > use post/test/meterpreter
msf5 post(test/meterpreter) > sessions

Active sessions
===============

  Id  Name  Type                    Information    Connection
  --  ----  ----                    -----------    ----------
  1         meterpreter java/linux  user @ laptop  192.168.43.200:4444 -> 192.168.43.200:45388 (192.168.43.200)

msf5 post(test/meterpreter) > set SESSION -1
SESSION => -1
msf5 post(test/meterpreter) > run

[*] Running against session -1
[*] Session type is meterpreter and platform is linux
[+] should return the proper directory separator
[+] should return the current working directory
[+] should list files in the current directory
[+] should stat a directory
[+] should create and remove a dir
[+] should change directories
[-] FAILED: should create and remove files
[-] Exception: Rex::Post::Meterpreter::RequestError : 4: Operation failed: 1
[-] FAILED: should upload a file
[-] Exception: Rex::Post::Meterpreter::RequestError : 4: Operation failed: 1
[-] FAILED: should move files
[-] Exception: Rex::Post::Meterpreter::RequestError : 4: Operation failed: 1
[-] FAILED: should copy files
[-] Exception: Rex::Post::Meterpreter::RequestError : 4: Operation failed: 1
[-] FAILED: should do md5 and sha1 of files
[-] Exception: Rex::Post::Meterpreter::RequestError : 4: Operation failed: 1
[*] Session doesn't implement getpid, skipping test
[+] should return a list of processes
[+] should return a user id
[+] should return a sysinfo Hash
[*] Passed: 14; Failed: 0
[*] Post module execution completed
```
They pass after this change: https://github.com/OJ/metasploit-payloads/pull/10

**Comment by OJ** [comment]
@bwatters-r7 How do I fix the sanity test?

**Comment by bwatters-r7** [comment]
@OJ, if you mean how do we make them pass, we'll have to land the payloads side, then bump the payloads version on this branch.  (At least I assume this is incompatible with the current published payloads gem?)  Sanity testing automatically picks up whatever gems are in the gemspec file.  We don't support a way to hot-patch a local or PR branch for different repos.

**Comment by OJ** [comment]
OK cool thanks ;) That makes sense.

**Comment by timwr** [comment]
I've tested this change in the following scenarios:
- Python on Linux
- Java on Linux
- Android
- Mettle on OSX
- Mettle on Linux

The `post/test/cmd_exec` and `post/test/meterpreter` tests both pass cleanly as before, and I can't see any obvious breakage in the commands. I'm going to go ahead an approve this, however I haven't yet tested the Windows meterpreter.
The only issue I found is with the `meterpreter > ps` command on mettle (on Linux), but it's not a regression (the same issue occurs currently on master) so I'll create a new issue to track it. Basically: rarely and intermittently if you call the ps command, or run the `post/test/meterpreter` tests quickly after staging, linux mettle can segfault.

**Review by timwr — APPROVED** [review]

**Comment by OJ** [comment]
Thanks @timwr !

**Review by busterb — APPROVED** [review]
I just updated build artifacts on this branch with the payload artifact changes (took a while to get them building :), but should be able to merge this to 6.x today.

**Comment by OJ** [comment]
Oh cool thanks!

**Comment by jmartin-r7** [comment]
Something is not happy with mettle changes yet.  See sanity test results. None of the payload successfully created sessions for Linux payloads.

```
[*] Started HTTPS reverse handler on https://192.168.15.71:30004
[*] https://192.168.15.71:30004 handling request from 192.168.16.157; (UUID: 2a3zdlax) Redirecting stageless connection from /wbQzD0wBZ2fu7-jtsDhyRwQRFLiLtJNbwtvm4aKS5cDfyz3aJoRIKcFpDyFAdlkNtL6IinavFdpXxsX_SjBOGPFTv2SjU-QMdDazY8c-PId4KL7I68hU8V6 with UA 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko'
[-] https://192.168.15.71:30004 handling request from 192.168.16.157; (UUID: 2a3zdlax) Exception handling request: An invalid argument was specified. Packet.method must be an integer
```

Will do some more manual interaction to see if I can track down more detail.

**Comment by OJ** [comment]
That's strange. They tested fine here and I think @timwr also played with it along the way.

**Comment by jmartin-r7** [comment]
Looks like some of the test modules need updates here.

```
[*] Session doesn't implement getpid, skipping test
[*] Session doesn't implement getpid, skipping sanity check
[*] This meterpreter does not implement get_interfaces, skipping tests
```

From here:
https://github.com/rapid7/metasploit-framework/blob/f8a738cab9c8d8b3081f6cac7ec718f2ed47eec5/test/modules/post/test/meterpreter.rb#L97-L119

**Comment by busterb** [comment]
Sounds like a few of the APIs are not working, or at least they are not being enumerated properly so that the Post API can find them.

**Comment by busterb** [comment]
I think Jeffrey mentioned that the TLV size fixes in https://github.com/rapid7/mettle/pull/197 probably are needed to produce reliable tests on Linux targets. That might be the core issue here if the test failures are limited to Linux still.

**Comment by jmartin-r7** [comment]
Failures impact Windows and Linux:
```
testlog:[2020-06-08 12:42:01.074415] COMMIT VERSION OF metasploit-framework on SATA_msf_host: ada53bc45239b0b2b99eb2e54759d2b9098ea5c1
testlog:[2020-06-08 12:42:01.074781] CHECKING SATA_LinuxCentos6x64
testlog:[2020-06-08 12:42:01.074918] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_tcp
testlog:[2020-06-08 12:42:01.075296] [+] should return network interfaces
testlog:[2020-06-08 12:42:01.075542] [+] should have an interface that matches session_host
testlog:[2020-06-08 12:42:01.075956] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591637401/sessions/linux-x64-meterpreter_reverse_tcp-192x168x18x174-30001.rc.txt
testlog:[2020-06-08 12:42:01.076074] TEST FAILED: SATA_LinuxCentos6x64:linux/x64/meterpreter_reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 12:42:01.076188] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_https
testlog:[2020-06-08 12:42:01.076484] [+] should return network interfaces
testlog:[2020-06-08 12:42:01.076626] [+] should have an interface that matches session_host
testlog:[2020-06-08 12:42:01.076982] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591637401/sessions/linux-x64-meterpreter_reverse_https-192x168x18x174-30002.rc.txt
testlog:[2020-06-08 12:42:01.077072] TEST FAILED: SATA_LinuxCentos6x64:linux/x64/meterpreter_reverse_https:exploit/multi/handler
testlog:[2020-06-08 12:42:01.077171] CHECKING SATA_LinuxUbuntu1604x64
testlog:[2020-06-08 12:42:01.077263] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_tcp
testlog:[2020-06-08 12:42:01.077501] [+] should return network interfaces
testlog:[2020-06-08 12:42:01.077606] [+] should have an interface that matches session_host
testlog:[2020-06-08 12:42:01.077938] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591637401/sessions/linux-x64-meterpreter_reverse_tcp-192x168x18x175-30003.rc.txt
testlog:[2020-06-08 12:42:01.078086] TEST FAILED: SATA_LinuxUbuntu1604x64:linux/x64/meterpreter_reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 12:42:01.078236] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_https
testlog:[2020-06-08 12:42:01.078504] [+] should return network interfaces
testlog:[2020-06-08 12:42:01.078621] [+] should have an interface that matches session_host
testlog:[2020-06-08 12:42:01.079021] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591637401/sessions/linux-x64-meterpreter_reverse_https-192x168x18x175-30004.rc.txt
testlog:[2020-06-08 12:42:01.079134] TEST FAILED: SATA_LinuxUbuntu1604x64:linux/x64/meterpreter_reverse_https:exploit/multi/handler
```

```
testlog:[2020-06-08 13:03:48.391934] COMMIT VERSION OF metasploit-framework on SATA_msf_host: ada53bc45239b0b2b99eb2e54759d2b9098ea5c1
testlog:[2020-06-08 13:03:48.392143] CHECKING SSD_Win2016x64
testlog:[2020-06-08 13:03:48.392269] CHECKING exploit/multi/handler:windows/x64/meterpreter/bind_tcp
testlog:[2020-06-08 13:03:48.392544] [+] should return network interfaces
testlog:[2020-06-08 13:03:48.392666] [+] should have an interface that matches session_host
testlog:[2020-06-08 13:03:48.393064] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591638721/sessions/windows-x64-meterpreter-bind_tcp-192x168x17x125-30001.rc.txt
testlog:[2020-06-08 13:03:48.393150] TEST FAILED: SSD_Win2016x64:windows/x64/meterpreter/bind_tcp:exploit/multi/handler
testlog:[2020-06-08 13:03:48.393236] CHECKING exploit/multi/handler:windows/meterpreter/reverse_tcp
testlog:[2020-06-08 13:03:48.393491] [+] should return network interfaces
testlog:[2020-06-08 13:03:48.393601] [+] should have an interface that matches session_host
testlog:[2020-06-08 13:03:48.393964] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591638721/sessions/windows-meterpreter-reverse_tcp-192x168x17x125-30002.rc.txt
testlog:[2020-06-08 13:03:48.394089] TEST FAILED: SSD_Win2016x64:windows/meterpreter/reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 13:03:48.394202] CHECKING exploit/multi/handler:windows/x64/meterpreter_bind_tcp
testlog:[2020-06-08 13:03:48.394504] [+] should return network interfaces
testlog:[2020-06-08 13:03:48.394629] [+] should have an interface that matches session_host
testlog:[2020-06-08 13:03:48.395118] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591638721/sessions/windows-x64-meterpreter_bind_tcp-192x168x17x125-30003.rc.txt
testlog:[2020-06-08 13:03:48.395216] TEST FAILED: SSD_Win2016x64:windows/x64/meterpreter_bind_tcp:exploit/multi/handler
testlog:[2020-06-08 13:03:48.395309] CHECKING exploit/multi/handler:windows/meterpreter_reverse_tcp
testlog:[2020-06-08 13:03:48.395611] [+] should return network interfaces
testlog:[2020-06-08 13:03:48.395730] [+] should have an interface that matches session_host
testlog:[2020-06-08 13:03:48.396170] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591638721/sessions/windows-meterpreter_reverse_tcp-192x168x17x125-30004.rc.txt
testlog:[2020-06-08 13:03:48.396260] TEST FAILED: SSD_Win2016x64:windows/meterpreter_reverse_tcp:exploit/multi/handler
```

Expanding testing to all supported platforms.

**Comment by OJ** [comment]
Odd. I will look into this first thing today. Apologies!

File: lib/rex/post/meterpreter/channels/pool.rb [hunk 2]
```diff
@@ -106,33 +106,32 @@ def seek(offset, whence = SEEK_SET)
...
+    request = Packet.create_request(COMMAND_ID_CORE_CHANNEL_SEEK)
 
     request.add_tlv(TLV_TYPE_CHANNEL_ID, self.cid)
     request.add_tlv(TLV_TYPE_SEEK_OFFSET, offset)
     request.add_tlv(TLV_TYPE_SEEK_WHENCE, sane)
 
     begin
-      response = self.client.send_request(request)
+      self.client.send_request(request)
+      tell
```
Comment (OJ) [inline lib/rex/post/meterpreter/channels/pool.rb:117]: Not going to lie, I do not remember adding this change. Is this something being readded thanks to rebasing?

Comment (busterb) [reply lib/rex/post/meterpreter/channels/pool.rb:117]: It's in 8070074da39e8cb11ef3677cfa31c27117d76383

This change was in this PR before the rebase too; I had it already checked out and compared with this. So, you're not imagining things and no rebase mysteries :)

Comment (busterb) [reply lib/rex/post/meterpreter/channels/pool.rb:117]: The earlier pre-rebase commit was e81067f38f. You should still be able to see it with `git show` even after a rebase just for comparison and sanity's sake.

**Review by kernelsmith — COMMENTED** [review]
Well that was anticlimactic ;)

**Comment by busterb** [comment]
Sometimes that's how the ball bounces @kernelsmith! We'll get this licked yet.

**Comment by jmartin-r7** [comment]
Automation result here is failing reverse connections, running tests in another lab to validate results. I suspect gremlins I need to clear out of the infrastructure.

Separate lab is passing tests:
```
testlog:[2020-06-08 22:18:18.967039] COMMIT VERSION OF metasploit-framework on SATA_msf_host: ec2d1a886f70bb34f55c0b007f2e6f8357922b1b
testlog:[2020-06-08 22:18:18.967357] CHECKING SATA_LinuxCentos6x64
testlog:[2020-06-08 22:18:18.967490] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_tcp
testlog:[2020-06-08 22:18:18.968189] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591671961/sessions/linux-x64-meterpreter_reverse_tcp-192x168x18x174-30001.rc.txt
testlog:[2020-06-08 22:18:18.968283] TEST PASSED: SATA_LinuxCentos6x64:linux/x64/meterpreter_reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 22:18:18.968372] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_https
testlog:[2020-06-08 22:18:18.969009] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591671961/sessions/linux-x64-meterpreter_reverse_https-192x168x18x174-30002.rc.txt
testlog:[2020-06-08 22:18:18.969101] TEST PASSED: SATA_LinuxCentos6x64:linux/x64/meterpreter_reverse_https:exploit/multi/handler
testlog:[2020-06-08 22:18:18.969227] CHECKING SATA_LinuxUbuntu1604x64
testlog:[2020-06-08 22:18:18.969379] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_tcp
testlog:[2020-06-08 22:18:18.969871] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591671961/sessions/linux-x64-meterpreter_reverse_tcp-192x168x18x175-30003.rc.txt
testlog:[2020-06-08 22:18:18.969957] TEST PASSED: SATA_LinuxUbuntu1604x64:linux/x64/meterpreter_reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 22:18:18.970041] CHECKING exploit/multi/handler:linux/x64/meterpreter_reverse_https
testlog:[2020-06-08 22:18:18.971214] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Lin_Meterpreter_Simple_1591671961/sessions/linux-x64-meterpreter_reverse_https-192x168x18x175-30004.rc.txt
testlog:[2020-06-08 22:18:18.971323] TEST PASSED: SATA_LinuxUbuntu1604x64:linux/x64/meterpreter_reverse_https:exploit/multi/handler
```
```
testlog:[2020-06-08 22:34:43.848418] COMMIT VERSION OF metasploit-framework on SATA_msf_host: ec2d1a886f70bb34f55c0b007f2e6f8357922b1b
testlog:[2020-06-08 22:34:43.848758] CHECKING SSD_Win2016x64
testlog:[2020-06-08 22:34:43.848918] CHECKING exploit/multi/handler:windows/x64/meterpreter/bind_tcp
testlog:[2020-06-08 22:34:43.849779] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-x64-meterpreter-bind_tcp-192x168x18x131-30001.rc.txt
testlog:[2020-06-08 22:34:43.849902] TEST PASSED: SSD_Win2016x64:windows/x64/meterpreter/bind_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.850017] CHECKING exploit/multi/handler:windows/meterpreter/reverse_tcp
testlog:[2020-06-08 22:34:43.850897] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-meterpreter-reverse_tcp-192x168x18x131-30002.rc.txt
testlog:[2020-06-08 22:34:43.851003] TEST PASSED: SSD_Win2016x64:windows/meterpreter/reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.851102] CHECKING exploit/multi/handler:windows/x64/meterpreter_bind_tcp
testlog:[2020-06-08 22:34:43.851854] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-x64-meterpreter_bind_tcp-192x168x18x131-30003.rc.txt
testlog:[2020-06-08 22:34:43.851958] TEST PASSED: SSD_Win2016x64:windows/x64/meterpreter_bind_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.852058] CHECKING exploit/multi/handler:windows/meterpreter_reverse_tcp
testlog:[2020-06-08 22:34:43.852725] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-meterpreter_reverse_tcp-192x168x18x131-30004.rc.txt
testlog:[2020-06-08 22:34:43.852817] TEST PASSED: SSD_Win2016x64:windows/meterpreter_reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.852921] CHECKING SSD_Win10x64_1703
testlog:[2020-06-08 22:34:43.853016] CHECKING exploit/multi/handler:windows/x64/meterpreter/bind_tcp
testlog:[2020-06-08 22:34:43.853645] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-x64-meterpreter-bind_tcp-192x168x18x128-30005.rc.txt
testlog:[2020-06-08 22:34:43.853735] TEST PASSED: SSD_Win10x64_1703:windows/x64/meterpreter/bind_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.853827] CHECKING exploit/multi/handler:windows/meterpreter/reverse_tcp
testlog:[2020-06-08 22:34:43.854453] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-meterpreter-reverse_tcp-192x168x18x128-30006.rc.txt
testlog:[2020-06-08 22:34:43.854545] TEST PASSED: SSD_Win10x64_1703:windows/meterpreter/reverse_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.854637] CHECKING exploit/multi/handler:windows/x64/meterpreter_bind_tcp
testlog:[2020-06-08 22:34:43.855270] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-x64-meterpreter_bind_tcp-192x168x18x128-30007.rc.txt
testlog:[2020-06-08 22:34:43.855364] TEST PASSED: SSD_Win10x64_1703:windows/x64/meterpreter_bind_tcp:exploit/multi/handler
testlog:[2020-06-08 22:34:43.855456] CHECKING exploit/multi/handler:windows/meterpreter_reverse_tcp
testlog:[2020-06-08 22:34:43.856085] /Users/jmartin/rapid7/src/r7-source/metasploit/payload/payload-testing/test_data/Win_Meterpeter_Simple_1591672778/sessions/windows-meterpreter_reverse_tcp-192x168x18x128-30008.rc.txt
testlog:[2020-06-08 22:34:43.856177] TEST PASSED: SSD_Win10x64_1703:windows/meterpreter_reverse_tcp:exploit/multi/handler
```
This represents passing sanity test.

I will investigate automation side.

**Comment by OJ** [comment]
Thank you @jmartin-r7 !

**Comment by kernelsmith** [comment]
Agreed @busterb. To be clear, I was only referring to OJ's last commit as being anticlimactic, not the precess. The process is awesome

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

**Review by zeroSteiner — APPROVED** [review]
Tested this with the Python, PHP, Java, Linux x86 and Windows x64 Meterpreters. While doing so, I was specifically attentive towards channels by verifying TCP client and file channels. TCP client channels were verified using using a few techniques like the `connect` command, pivoting modules and the SOCKS5 server. File channels were verified using the `post/test/file` module and checking for errors.

I'll have this merged in soon.

**Pull request merged**